core/chain_indexer.go

```diff
@@ -473,3 +512,121 @@ func (c *ChainIndexer) removeSectionHead(section uint64) {
 
 	c.indexDb.Delete(append([]byte("shead"), data[:]...))
 }
+
+func (c *ChainIndexer) addOutpointsToIndexer(addressOutpoints map[string][]*types.OutpointAndDenomination, nodeCtx int, config params.ChainConfig, block *types.WorkObject) {
+	utxos := block.QiTransactions()
+
+	for _, tx := range utxos {
+		for _, in := range tx.TxIn() {
+
+			if types.IsCoinBaseTx(tx, block.ParentHash(nodeCtx), config.Location) {
+				// skip coinbase
+				continue
+			}
+
+			outpoint := in.PreviousOutPoint
+
+			addr := crypto.PubkeyBytesToAddress(in.PubKey, config.Location).Hex()
+			outpointsForAddress := addressOutpoints[addr]
+
+			for i := 0; i < len(outpointsForAddress); i++ {
+				if &outpointsForAddress[i].TxHash == &outpoint.TxHash && outpointsForAddress[i].Index == outpoint.Index {
+					outpointsForAddress = append(outpointsForAddress[:i], outpointsForAddress[i+1:]...)
+					addressOutpoints[addr] = outpointsForAddress
+
+					break
+				}
+			}
+		}
+
+		for i, out := range tx.TxOut() {
+
+			address := out.Address
+			outpoint := types.OutPoint{
+				TxHash: tx.Hash(),
+				Index:  uint16(i),
+			}
+
+			addr := common.BytesToAddress(address, config.Location).Hex()
+
+			outpointAndDenom := &types.OutpointAndDenomination{
+				TxHash:       outpoint.TxHash,
+				Index:        outpoint.Index,
+				Denomination: out.Denomination,
+			}
+
+			if existingOutpoints, ok := addressOutpoints[addr]; ok {
+				addressOutpoints[addr] = append(existingOutpoints, outpointAndDenom)
+			} else {
+				addressOutpoints[addr] = []*types.OutpointAndDenomination{outpointAndDenom}
+			}
+		}
+	}
+}
+
+func (c *ChainIndexer) reorgUtxoIndexer(headers []*types.WorkObject, addressOutpoints map[string][]*types.OutpointAndDenomination, nodeCtx int, config params.ChainConfig) error {
+	for _, header := range headers {
+		block := rawdb.ReadWorkObject(c.chainDb, header.Hash(), types.BlockObject)
+
+		for _, tx := range block.QiTransactions() {
+			for i, out := range tx.TxOut() {
+
+				address := out.Address
+
+				outpoint := types.OutpointAndDenomination{
+					TxHash:       tx.Hash(),
+					Index:        uint16(i),
+					Denomination: out.Denomination,
+				}
+
+				addr := common.BytesToAddress(address, config.Location).Hex()
+				outpointsForAddress := addressOutpoints[addr]
+
+				for i := 0; i < len(outpointsForAddress); i++ {
+					if outpointsForAddress[i].TxHash == outpoint.TxHash && outpointsForAddress[i].Index == outpoint.Index {
```

Comment:
Why are we using an array instead of a map? This indexing logic is easy to mess up and way less ergonomic.

Reply:
Are you suggesting `addressOutpoints map[string][]*types.OutpointAndDenomination` should be `addressOutpoints map[string]map[TxHash + index]*types.OutpointAndDenomination`